include/rocksdb/customizable.h

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#pragma once
+
+#include "rocksdb/configurable.h"
+#include "rocksdb/status.h"
+
+namespace ROCKSDB_NAMESPACE {
+/**
+ * Customizable a base class used by the rocksdb that describes a
+ * standard way of configuring and creating objects.  Customizable objects
+ * are configurable objects that can be created from an ObjectRegistry.
+ *
+ * When a Customizable is being created, the "name" property specifies
+ * the name of the instance being created.
+ * For custom objects, their configuration and name can be specified by:
+ * [prop]={name=X;option 1 = value1[; option2=value2...]}
+ *
+ * [prop].name=X
+ * [prop].option1 = value1
+ *
+ * [prop].name=X
+ * X.option1 =value1
+ */
+class Customizable : public Configurable {
```

Comment:
Can you explain the essentialness of the customizable class. Why can't we directly use Configurable? What's the main differences?

Reply:
Almost every class that is Configurable is also Customizable.   A few that are not are the wrappers that deal with DBOptions and ColumnFamilyOptions.  Another is the OptionsRegistry itself.

When there are a choice of implementations for something (like different Cache or Table implementations), those are Customizable.  When there are choices, each choice has a Name() and there is a factory to create these by name (CreateFromString).

When there is a single implementation class for the options, that is Configurable.  There is no name and no factory, as there is only one choice of things to create.

A Configurable is very similar to a "struct with options" (like CompressionOptions and its associated map of OptionTypeInfo). 

Customizable implies multiple implementations; Configurable implies a single implementation.

I added a comment documenting this but it will be a pretty odd comment.  The "base class" (e.g. Table, Cache) that extends a Customizable is expected to implement certain methods (Type, CreateFromString).  The "implementation classes" are expected to implement certain methods (Name).